crates/ide-assists/src/handlers/convert_nested_function_to_closure.rs

```diff
@@ -0,0 +1,147 @@
+use hir::Semantics;
+use ide_db::assists::{AssistId, AssistKind};
+use ide_db::base_db::FileId;
+use ide_db::RootDatabase;
+use syntax::ast::{self, HasName};
+use syntax::{AstNode, SyntaxKind};
+
+use crate::assist_context::{AssistContext, Assists};
+
+// Assist: convert_nested_function_to_closure
+//
+// Converts a function that is defined within the body of another function into a closure.
+//
+// ```
+// fn main() {
+//     fn foo(label: &str, $0number: u64) {
+//         println!("{}: {}", label, number);
+//     }
+//
+//     foo("Bar", 100);
+// }
+// ```
+// ->
+// ```
+// fn main() {
+//     let foo = |label: &str, number: u64| {
+//         println!("{}: {}", label, number);
+//     };
+//
+//     foo("Bar", 100);
+// }
+// ```
+pub(crate) fn convert_nested_function_to_closure(
+    acc: &mut Assists,
+    ctx: &AssistContext<'_>,
+) -> Option<()> {
+    let function = ctx.find_node_at_offset::<ast::Fn>()?;
+    if !is_nested_function(&function) {
+        return None;
+    }
+
+    let target = function.syntax().text_range();
+    let body = function.body()?;
+    let name = function.name()?;
+    let params = function.param_list()?;
+    let has_semicolon = has_semicolon(&function, ctx.file_id(), &ctx.sema);
+
+    acc.add(
+        AssistId("convert_nested_function_to_closure", AssistKind::RefactorRewrite),
+        "Convert nested function to closure",
+        target,
+        |edit| {
+            let params_text = params.syntax().text().to_string();
+            let params_text_trimmed =
+                params_text.strip_prefix("(").and_then(|p| p.strip_suffix(")"));
+
+            if let Some(closure_params) = params_text_trimmed {
+                let body = body.to_string();
+                let body = if has_semicolon { body } else { format!("{};", body) };
+                edit.replace(target, format!("let {} = |{}| {}", name, closure_params, body));
+            }
+        },
+    )
+}
+
+/// Returns whether the given function is nested within the body of another function.
+fn is_nested_function(function: &ast::Fn) -> bool {
+    function
+        .syntax()
+        .parent()
+        .map(|p| p.ancestors().any(|a| a.kind() == SyntaxKind::FN))
+        .unwrap_or(false)
```

Comment:
This will trigger incorrectly if we have an associated function nested in a function (very rare that someone does this but it could happen), so it might be nicer to do the following
```rs
function.ancestors().skip(1).find_map(ast::Item::cast).unwrap_or(false, |it| matches!(it, ast::item::Fn(_))
```

Reply:
I'm not sure what that snippet is supposed to do, namely because there are two arguments that are passed to `unwrap_or`. Would it be easier to just check if the function is an associated function?

Reply:
oh my bad, this was supposed to be a `map_or_else`. But yes it would be easier to just check if it is an assocaited function actually, that is go from the `ast::Fn` to the `hir` version and check if that is an assoc item.